crates/compose/src/lib.rs

```diff
@@ -0,0 +1,406 @@
+            for import_name in matched.into_iter() {
+                if let Some(previous_info) = prepared.insert(import_name.clone(), info.clone()) {
+                    anyhow::bail!(
+                        "component '{component_id}' dependency '{dependency_name}' introduces import '{import_name}' which conflicts with dependency '{previous_dependency_name}'",
+                        component_id = component.id,
+                        previous_dependency_name = previous_info.manifest_name,
+                    );
+                }
+            }
+        }
+
+        Ok(prepared)
+    }
+
+    async fn build_instantiation_arguments(
+        &mut self,
+        world_id: WorldId,
+        dependencies: IndexMap<String, DependencyInfo>,
+    ) -> anyhow::Result<IndexMap<String, NodeId>> {
+        let mut cache = Default::default();
+        let mut checker = SubtypeChecker::new(&mut cache);
+
+        let mut arguments = IndexMap::new();
+
+        for (import_name, dependency_info) in dependencies {
+            let (export_name, export_ty) = match dependency_info.export_name {
+                Some(export_name) => {
+                    let Some(export_ty) = self.graph.types()[dependency_info.world_id]
+                        .exports
+                        .get(&export_name)
+                    else {
+                        anyhow::bail!(
+                                "dependency '{dependency_name}' doesn't export '{export_name}' to satisfy import '{import_name}'",
+                                dependency_name = dependency_info.manifest_name,
+                            );
+                    };
+
+                    (export_name, export_ty)
+                }
+                None => {
+                    let Some(export_ty) = self.graph.types()[dependency_info.world_id]
+                        .exports
+                        .get(&import_name)
+                    else {
+                        anyhow::bail!(
+                                "dependency '{dependency_name}' doesn't export '{import_name}' to satisfy import '{import_name}'",
+                                dependency_name = dependency_info.manifest_name,
+                            );
+                    };
+
+                    (import_name.clone(), export_ty)
+                }
+            };
+
+            let import_ty = self.graph.types()[world_id]
+                .imports
+                .get(&import_name)
+                .unwrap();
+
+            // Ensure that export_ty is a subtype of import_ty
+            checker.is_subtype(
+                *export_ty,
+                self.graph.types(),
+                *import_ty,
+                self.graph.types(),
+            ).with_context(|| {
+                format!(
+                    "dependency '{dependency_name}' export '{export_name}' of type '{export_ty:?}' is not a subtype of import '{import_name}' of type '{import_ty:?}'",
```

Comment:
`export_ty` and `import_ty` are meaningless in debug form (e.g., they look like `Instance(InterfaceId(Id { idx: 31 }))`). 

We might be better off completely removing the type information as it might be more confusing than helpful. For example, do we expect users to know what an instance is? 

If we keep the type information we should at the very least do:

```rust
                let export_ty = export_ty.desc(&self.graph.types());
                let import_ty = import_ty.desc(&self.graph.types());
```

But I would do:

```suggestion
                    "dependency '{dependency_name}' exports '{export_name}' which is not compatible with import '{import_name}'",
```